Newsletters: add 276 (2023-11-08) (#1384)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2023-11-08-newsletter.md

  balance could be refunded to the other party.  Halseth notes that this
  would be more efficient onchain and could make it more difficult to
  perfrom certain types of [channel jamming attacks][topic channel
  jamming attacks].

## Bitcoin Core PR Review Club

*In this monthly section, we summarize a recent [Bitcoin Core PR Review
Club][] meeting, highlighting some of the important questions and
answers.  Click on a question below to see a summary of the answer from
the meeting.*

[Fee Estimator updates from Validation Interface/CScheduler thread][review club 28368]
is a PR by Abubakar Sadiq Ismail (ismaelsadeeq)
that modifies the way the transaction fee estimator integrates with the
rest of the `bitcoind` client.
It moves fee estimator updates from occurring synchronously during mempool
updates (transactions being added or removed) to instead occur asynchronously.
This is a performance improvement.

When a new block is found, its transactions
to be removed from the mempool.
Since block validation and relay are performance-critical, it's beneficial
to reduce the required amount of work -- such as updating the fee estimator.

> {% include functions/details-list.md

Reviewer: details-list.md can handle 10 Q&A's as an upper bound. I think its probably best either way to trim the number of questions highlighted this week.

IMHO q3,q5,q6 could be candidates for trimming, but Im not deeply familiar with the PR and review

Reviewer 2: I agree q3 and q5 are good candidates for trimming. Would also suggest q11.

    forum for interesting or important discussions effective
    immediately.

- **HTLC aggregation with covenants:** Johan Torås Halseth
  [posted][halseth agg] to the Lightning-Dev mailing list a suggestion
  for using a [covenant][topic covenants] to aggregate multiple
  [HTLCs][topic htlc] into a single output that could be spent all at
  once if a party knew all the preimages.  If a party only knew some of
  the preimages, they could claim just those and then the remaining
  balance could be refunded to the other party.  Halseth notes that this
  would be more efficient onchain and could make it more difficult to
  perfrom certain types of [channel jamming attacks][topic channel
  jamming attacks].

## Bitcoin Core PR Review Club

*In this monthly section, we summarize a recent [Bitcoin Core PR Review
Club][] meeting, highlighting some of the important questions and
answers.  Click on a question below to see a summary of the answer from
the meeting.*

[Fee Estimator updates from Validation Interface/CScheduler thread][review club 28368]
is a PR by Abubakar Sadiq Ismail (ismaelsadeeq)
that modifies the way the transaction fee estimator data is updated.

> (Fee estimation is used with the node's owner initiates a transaction.)

Reviewer 3: ```suggestion
(Fee estimation is used when the node's owner initiates a transaction.)
```

---
title: 'Bitcoin Optech Newsletter #276'
permalink: /en/newsletters/2023/11/08/
name: 2023-11-08-newsletter
slug: 2023-11-08-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME

## News

> - **Discontinuation of mailing lists:** administrators for the

Reviewer 4: There will be a forthcoming announcement as to where the bitcoin-dev mailing list might be moving to; could this be reworded as something less than total death?

Reviewer 4: Thank you, looks better now.

  for using a [covenant][topic covenants] to aggregate multiple
  [HTLCs][topic htlc] into a single output that could be spent all at
  once if a party knew all the preimages.  If a party only knew some of
  the preimages, they could claim just those and then the remaining
  balance could be refunded to the other party.  Halseth notes that this
  would be more efficient onchain and could make it more difficult to
  perfrom certain types of [channel jamming attacks][topic channel
  jamming attacks].

## Bitcoin Core PR Review Club

*In this monthly section, we summarize a recent [Bitcoin Core PR Review
Club][] meeting, highlighting some of the important questions and
answers.  Click on a question below to see a summary of the answer from
the meeting.*

[Fee Estimator updates from Validation Interface/CScheduler thread][review club 28368]
is a PR by Abubakar Sadiq Ismail (ismaelsadeeq)
that modifies the way the transaction fee estimator integrates with the
rest of the `bitcoind` client.
It moves fee estimator updates from occurring synchronously during mempool
updates (transactions being added or removed) to instead occur asynchronously.

> This is a performance improvement.

Reviewer 5: Great suggestion, I forgot about that. I decided to add it to be part of the answer to the first question, since you mentioned this during the chat, rather than in the notes. (I think of this part of the summary as mainly summarizing the notes rather than the discussion.) But let me know if you feel strongly the other way.

When a new block is found, its transactions

> to be removed from the mempool.

Reviewer 5: Good catch, what I wrote isn't even English!

  for using a [covenant][topic covenants] to aggregate multiple
  [HTLCs][topic htlc] into a single output that could be spent all at
  once if a party knew all the preimages.  If a party only knew some of
  the preimages, they could claim just those and then the remaining
  balance could be refunded to the other party.  Halseth notes that this
  would be more efficient onchain and could make it more difficult to
  perfrom certain types of [channel jamming attacks][topic channel
  jamming attacks].

## Bitcoin Core PR Review Club

*In this monthly section, we summarize a recent [Bitcoin Core PR Review
Club][] meeting, highlighting some of the important questions and
answers.  Click on a question below to see a summary of the answer from
the meeting.*

[Fee Estimator updates from Validation Interface/CScheduler thread][review club 28368]
is a PR by Abubakar Sadiq Ismail (ismaelsadeeq)
that modifies the way the transaction fee estimator integrates with the
rest of the `bitcoind` client.
It moves fee estimator updates from occurring synchronously during mempool
updates (transactions being added or removed) to instead occur asynchronously.

> This is a performance improvement.

Reviewer 2: Well, not really. I imagine this is actually slower overall, but we're no longer blocking the really important task of processing blocks with it.

Reviewer 5: Ah, I see the distinction, you're right, more CPU cycles will be expended almost for sure, but it's the critical path that's faster. I'll improve this wording.

{% include functions/details-list.md
  q0="Why is it beneficial to remove `CTxMempool`'s dependency on `CBlockPolicyEstimator`?"
  a0="Currently, upon receiving a new block, its validation is blocked while
      fee estimates, which depend on the new block, are updated. This delays
      the completion of validation, and also delays relaying the block to peers.
      Removing the `CTxMempool` dependency on `CBlockPolicyEstimator`
      allows fee estimates to be updated asynchronously (in a different thread)
      so that validation and relay can complete more quickly.
      It may also make testing `CTxMempool` easier."
  a0link="https://bitcoincore.reviews/28368#l-30"

  q1="Isn't fee estimation currently updated synchronously when transactions
      are added to or removed from the mempool even without a new block arriving?"
  a1="Yes, but performance isn't as critical at those times as during
      block validation and relay."
  a1link="https://bitcoincore.reviews/28368#l-41"

  q2="Are there any benefits of the `CBlockPolicyEstimator` being a member
      of `CTxMempool` and updating it synchronously (the current arrangement)?
      Are there downsides to removing it?"

>   a2="Synchronous code is simpler and easier to reason about. Also, the fee
>       estimator has more visibility into the entire mempool; a downside is
>       the need to encapsulate all the information needed for fee estimation
>       into a new `TransactionInfo` structure.
>       However, not much information is needed."

Reviewer 6: ```suggestion
  a2="Synchronous code is simpler and easier to reason about. Also, the fee
      estimator has more visibility into the entire mempool; a downside is
      the need to encapsulate all the information needed for fee estimation
      into a new `NewMempoolTransactionInfo` structure.
      However, not much information is needed by the fee estimator."
```

      any implemented `CValidationInterface` method will be executed each time
      the method callback is fired using `CMainSignals`.
      Callbacks are fired whenever the corresponding event occurs."
  a6link="https://bitcoincore.reviews/28368#l-90"

  q7="[`BlockConnected`][BlockConnected] and [`NewPoWValidBlock`][NewPoWValidBlock]
      are different callbacks.
      Which one is asynchronous and which one is synchronous? How can you tell?"
  a7="`BlockConnected` is asynchronous; `NewPoWValidBlock`is asynchronous.
      Asynchronous callbacks queue an \"event\" to be run later within the
      `CScheduler` thread."
  a7link="https://bitcoincore.reviews/28368#l-105"

  q8="In [commit 4986edb][], why are we adding a new callback,
      `MempoolTransactionsRemovedForConnectedBlock`, instead of using
      `BlockConnected` (which also indicates a transaction being removed
      from the mempool)?"
  a8="The fee estimator needs to know when transactions are removed from the
      mempool for any reason, not just when a block is connected.
      Also, the fee estimator needs a transaction's base fee, and this isn't

>       provided via `MempoolTransactionsRemovedForConnectedBlock` (which provides a `CBlock`).

Reviewer 2: Do you mean
```suggestion
      provided via `BlockConnected` (which provides a `CBlock`).
```
?

      We could add the base fee to the `block.vtx` (transaction list) entries,
      but it's undesirable to change such an important and ubiquitous data
      structure just to support fee estimation."
  a8link="https://bitcoincore.reviews/28368#l-130"

  q9="Why don’t we use a `std::vector<CTxMempoolEntry>` as a parameter of
      `MempoolTransactionsRemovedForBlock` callback?
      This would eliminate the requirement for a new struct type to hold the
      per-transaction information needed for fee estimation."
  a9="That wouldn't be efficient as there may be several thousand transactions
      in a block, and copying each one's `CTxMempoolEntry` would be expensive,

>       especially since each entry contains references to the transactions parents

Reviewer: ```suggestion
      especially since each entry contains references to the transaction's parents
```

      and children (which are not needed for fee estimation)."
  a9link="https://bitcoincore.reviews/28368#l-159"

%}
{% comment %}
  q10="How is the base fee of a `CTransactionRef` computed?"
  a10="It's the sum of the input values minus the sum of the output values.
      However, the callback can't access the input values because they're stored
      in the previous transaction outputs (which the callback doesn't have access to).

>       That's why the base fee is included in the `TransactionInfo` structure."

Reviewer 5: Not sure I understand this suggestion. The `TransactionInfo` struct has a member `m_fee`, but `NewMempoolTransactionInfo` only adds four `bool` members. (I'm looking at the latest commit.) Do you still think it's better to refer here to `NewMempoolTransactionInfo`?

      of each other.
      There doesn't seem to be much practical benefit to splitting.
      The current PR is more narrow, and minimally scoped to making fee
      estimates update asynchronously."
  a3link="https://bitcoincore.reviews/28368#l-71"

  q4="In a subclass, why is implementing a `CValidationInterface` method
      equivalent to subscribing to the event?"
  a4="All subclasses of  `CValidationInterface` are clients.
      The subclass can implement some or all `CValidationInterface` methods (callbacks),
      for example, connecting or disconnecting a block, or a transaction
      [added][tx add] to or [removed][tx remove] from the mempool.
      After being registered (by calling `RegisterSharedValidationInterface()`),
      any implemented `CValidationInterface` method will be executed each time
      the method callback is fired using `CMainSignals`.
      Callbacks are fired whenever the corresponding event occurs."
  a4link="https://bitcoincore.reviews/28368#l-90"

  q5="[`BlockConnected`][BlockConnected] and [`NewPoWValidBlock`][NewPoWValidBlock]
      are different callbacks.
      Which one is asynchronous and which one is synchronous? How can you tell?"

>   a5="`BlockConnected` is asynchronous; `NewPoWValidBlock`is asynchronous.

Reviewer 5: ```suggestion
  a5="`BlockConnected` is asynchronous; `NewPoWValidBlock` is asynchronous.
```

>   a5="`BlockConnected` is asynchronous; `NewPoWValidBlock` is asynchronous.

Reviewer 7: Isn't `NewPoWValidBlock` synchronous rather than asynchronous?

Reviewer: Took this suggestion

      This would eliminate the requirement for a new struct type to hold the
      per-transaction information needed for fee estimation."
  a7="The fee estimator doesn't need all of the fields from `CTxMempoolEntry`."
  a7link="https://bitcoincore.reviews/28368#l-159"

  q8="How is the base fee of a `CTransactionRef` computed?"
  a8="It's the sum of the input values minus the sum of the output values.
      However, the callback can't access the input values because they're stored
      in the previous transaction outputs (which the callback doesn't have access to).
      That's why the base fee is included in the `TransactionInfo` structure."
  a8link="https://bitcoincore.reviews/28368#l-166"
%}

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [Bitcoin Core 26.0rc2][] is a release candidate for the next major
  version of the predominant full node implementation.  Verified test
  binaries have not yet been released as of this writing, although we
  expect them to be published at the preceding URL shortly after

>   publication of the newsletter.  There's a brief overview to [suggested

Reviewer 2: rc2 binaries are up

Reviewer: Thanks @glozow just updated the text accordingly.

Proposals (BIPs)][bips repo], [Lightning BOLTs][bolts repo], and
[Bitcoin Inquisition][bitcoin inquisition repo].*

- [Core Lightning #6824][] updates the implementation of the
  [interactive funding protocol][topic dual funding] to "store state
  when sending `commitment_signed`, and [add] a `next_funding_txid`
  field to `channel_reestablish` to ask our peer to retransmit
  signatures that we haven't received."  It is based on an
  [update][36c04c8ac] to the proposed [dual funding PR][bolts #851].

- [Core Lightning #6783][] deprecates the `large-channels` configuration
  option, making [large channels][topic large channels] and large
  payment amounts always enabled.

- [Core Lightning #6780][] improves support for fee bumping onchain transactions
  associated with [anchor outputs][topic anchor outputs].

- [Core Lightning #6773][] allows the `decode` RPC to verify that the
  contents of a backup file are valid and contain the latest information
  necessary to perform a full recovery.

- [Core Lightning #6734][] updates the `listfunds` RPC to provide
  information users need if they want to [CPFP][topic cpfp] fee bump a
  channel mutual close transaction.

> - [Eclair #2761][] allows forwarding a limited number [HTLCs][topic

Reviewer 8: ```suggestion
- [Eclair #2761][] allows forwarding a limited number of [HTLCs][topic
```

Reviewer 8: not fixed
